node/src/components/chain_synchronizer.rs

```diff
@@ -363,92 +411,191 @@ impl ChainSynchronizer {
+        let maybe_era_end = maybe_switch_block_header_before_upgrade
             .as_ref()
-            .unwrap_or_else(|| block_and_execution_effects.block.header())
+            .unwrap_or_else(|| immediate_switch_block_and_exec_effects.block.header())
             .era_end();
+
         let validators_to_sign_immediate_switch_block = match maybe_era_end {
             Some(era_end) => era_end
                 .next_era_validator_weights()
                 .keys()
                 .cloned()
                 .collect(),
             None => {
-                error!(
-                    ?maybe_upgrade_block_header,
-                    "upgrade/genesis block is not a switch block"
-                );
+                error!("upgrade/genesis switch block missing era end");
                 return fatal!(
                     effect_builder,
-                    "upgrade/genesis block is not a switch block"
+                    "upgrade/genesis switch block missing era end"
                 )
                 .ignore();
             }
         };
+
+        // For an emergency upgrade, we always execute/commit locally rather than syncing over it.
+        // This means we should try fast syncing again if this was an emergency upgrade.
+        if is_emergency_upgrade {
+            return operations::run_fast_sync_task(
+                effect_builder,
+                self.config.clone(),
+                self.metrics.clone(),
+            )
+            .event(|result| Event::FastSyncAfterEmergencyUpgradeResult {
+                immediate_switch_block_and_exec_effects,
+                validators_to_sign_immediate_switch_block,
+                result,
+            });
+        }
+
+        let highest_block_header = immediate_switch_block_and_exec_effects
+            .block
+            .header()
+            .clone();
         self.joining_outcome = Some(JoiningOutcome::RanUpgradeOrGenesis {
-            block_and_execution_effects,
+            block_and_execution_effects: immediate_switch_block_and_exec_effects,
             validators_to_sign_immediate_switch_block,
+            highest_block_header,
         });
         Effects::new()
     }
 
+    fn handle_fast_sync_after_emergency_upgrade_result(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        immediate_switch_block_and_exec_effects: BlockAndExecutionEffects,
+        validators_to_sign_immediate_switch_block: HashSet<PublicKey>,
+        result: Result<FastSyncOutcome, Error>,
+    ) -> Effects<Event> {
+        match result {
+            Ok(FastSyncOutcome::ShouldCommitGenesis) => {
+                let msg = "fast sync after emergency upgrade should not require commit genesis";
+                error!(msg);
+                fatal!(effect_builder, "{}", msg).ignore()
+            }
+            Ok(FastSyncOutcome::ShouldCommitUpgrade { .. }) => {
+                let msg = "fast sync after emergency upgrade should not require commit upgrade";
```

Comment:
That case shouldn't be supported since we can't execute any blocks for a previous protocol version.  It's checked in `get_trusted_key_block_info`:
```rust
match ctx.config.last_emergency_restart() {
    Some(last_emergency_restart)
        if last_emergency_restart > current_header_to_walk_back_from.era_id()
            && !ctx.trusted_block_is_last_before_activation() =>
    {
        return Err(Error::TrustedHeaderEraTooEarly {
            trusted_header: Box::new(ctx.trusted_block_header().clone()),
            maybe_last_emergency_restart_era_id: ctx.config.last_emergency_restart(),
        })
    }
    _ => {}
}
```